controllers/vspheremachinetemplate_controller.go

```diff
@@ -0,0 +1,134 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	vmoprv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	vmwarev1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
+	capvcontext "sigs.k8s.io/cluster-api-provider-vsphere/pkg/context"
+)
+
+// +kubebuilder:rbac:groups=vmware.infrastructure.cluster.x-k8s.io,resources=vspheremachinetemplates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=vmware.infrastructure.cluster.x-k8s.io,resources=vspheremachinetemplates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachineclasses,verbs=get;list;watch
+
+// AddMachineTemplateControllerToManager adds the machine template controller to the provided
+// manager.
+func AddMachineTemplateControllerToManager(ctx context.Context, controllerManagerContext *capvcontext.ControllerManagerContext, mgr manager.Manager, supervisorBased bool, options controller.Options) error {
+	r := &machineTemplateReconciler{
+		Client:          controllerManagerContext.Client,
+		Recorder:        mgr.GetEventRecorderFor("vspheremachinetemplate-controller"),
+		supervisorBased: supervisorBased,
+	}
+
+	if supervisorBased {
+		return ctrl.NewControllerManagedBy(mgr).
+			// Watch the controlled, infrastructure resource.
+			For(&vmwarev1.VSphereMachineTemplate{}).
+			WithOptions(options).
+			// Watch the CAPI resource that owns this infrastructure resource.
+			Watches(
+				&vmoprv1.VirtualMachineClass{},
+				handler.EnqueueRequestsFromMapFunc(r.enqueueVirtualMachineClassToMachineTemplateRequests),
+			).
+			WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), controllerManagerContext.WatchFilterValue)).
+			Complete(r)
+	}
+
+	return nil
+}
+
+type machineTemplateReconciler struct {
+	Client          client.Client
+	Recorder        record.EventRecorder
+	supervisorBased bool
+}
+
+// Reconcile ensures the back-end state reflects the Kubernetes resource state intent.
```

Comment:
Can we drop large parts of this comment? It's super confusing (at least to me)

Reply:
Removed the whole